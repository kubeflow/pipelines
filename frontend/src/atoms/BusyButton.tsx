/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Button, { ButtonProps } from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';
import { stylesheet, classes } from 'typestyle';

const css = stylesheet({
  icon: {
    height: 20,
    marginRight: 4,
    width: 20,
  },
  root: {
    cursor: 'pointer',
    marginBottom: 2, // To prevent container from flickering as the spinner shows up
    position: 'relative',
    transition: 'padding 0.3s',
  },
  rootBusy: {
    cursor: 'default',
    paddingRight: 35,
  },
  spinner: {
    opacity: 0,
    position: 'absolute',
    right: '0.8em',
    transition: 'all 0.3s',
  },
  spinnerBusy: {
    opacity: 1,
  },
});

interface BusyButtonProps extends ButtonProps {
  title: string;
  icon?: any;
  busy?: boolean;
  outlined?: boolean;
}

class BusyButton extends React.Component<BusyButtonProps> {
  public render(): JSX.Element {
    const { title, busy, className, disabled, icon, outlined, ...rest } = this.props;

    return (
      <Button
        {...rest}
        color={outlined ? 'primary' : 'secondary'}
        className={classes(css.root, busy && css.rootBusy, className)}
        disabled={busy || disabled}
      >
        {!!icon && <this.props.icon className={css.icon} />}
        <span>{title}</span>
        {busy === true && (
          <CircularProgress size={15} className={classes(css.spinner, busy && css.spinnerBusy)} />
        )}
      </Button>
    );
  }
}

export default BusyButton;
