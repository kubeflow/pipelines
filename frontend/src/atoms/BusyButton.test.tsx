/*
 * Copyright 2018 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import BusyButton from './BusyButton';
import TestIcon from '@material-ui/icons/Help';
import { create } from 'react-test-renderer';

describe('BusyButton', () => {
  it('renders with just a title', () => {
    const tree = create(<BusyButton title='test busy button' />);
    expect(tree).toMatchSnapshot();
  });

  it('renders with a title and icon', () => {
    const tree = create(<BusyButton title='test busy button' icon={TestIcon} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders with a title and icon, and busy', () => {
    const tree = create(<BusyButton title='test busy button' icon={TestIcon} busy={true} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders disabled', () => {
    const tree = create(<BusyButton title='test busy button' icon={TestIcon} disabled={true} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders a primary outlined buton', () => {
    const tree = create(<BusyButton title='test busy button' outlined={true} />);
    expect(tree).toMatchSnapshot();
  });
});
