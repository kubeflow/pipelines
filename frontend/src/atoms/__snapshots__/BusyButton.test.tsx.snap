// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusyButton renders a primary outlined buton 1`] = `
<button
  className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-textPrimary-4 MuiButton-flat-6 MuiButton-flatPrimary-7 root"
  disabled={false}
  onBlur={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiButton-label-2"
  >
    <span>
      test busy button
    </span>
  </span>
  <span
    className="MuiTouchRipple-root-30"
  />
</button>
`;

exports[`BusyButton renders disabled 1`] = `
<button
  className="MuiButtonBase-root-27 MuiButtonBase-disabled-28 MuiButton-root-1 MuiButton-text-3 MuiButton-textSecondary-5 MuiButton-flat-6 MuiButton-flatSecondary-8 MuiButton-disabled-21 root"
  disabled={true}
  onBlur={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="-1"
  type="button"
>
  <span
    className="MuiButton-label-2"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-37 icon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
      />
    </svg>
    <span>
      test busy button
    </span>
  </span>
</button>
`;

exports[`BusyButton renders with a title and icon 1`] = `
<button
  className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-textSecondary-5 MuiButton-flat-6 MuiButton-flatSecondary-8 root"
  disabled={false}
  onBlur={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiButton-label-2"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-37 icon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
      />
    </svg>
    <span>
      test busy button
    </span>
  </span>
  <span
    className="MuiTouchRipple-root-30"
  />
</button>
`;

exports[`BusyButton renders with a title and icon, and busy 1`] = `
<button
  className="MuiButtonBase-root-27 MuiButtonBase-disabled-28 MuiButton-root-1 MuiButton-text-3 MuiButton-textSecondary-5 MuiButton-flat-6 MuiButton-flatSecondary-8 MuiButton-disabled-21 root rootBusy"
  disabled={true}
  onBlur={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="-1"
  type="button"
>
  <span
    className="MuiButton-label-2"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-37 icon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
      />
    </svg>
    <span>
      test busy button
    </span>
    <div
      className="MuiCircularProgress-root-46 MuiCircularProgress-colorPrimary-49 MuiCircularProgress-indeterminate-48 spinner spinnerBusy"
      role="progressbar"
      style={
        Object {
          "height": 15,
          "width": 15,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg-51"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle-52 MuiCircularProgress-circleIndeterminate-54"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </div>
  </span>
</button>
`;

exports[`BusyButton renders with just a title 1`] = `
<button
  className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-textSecondary-5 MuiButton-flat-6 MuiButton-flatSecondary-8 root"
  disabled={false}
  onBlur={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiButton-label-2"
  >
    <span>
      test busy button
    </span>
  </span>
  <span
    className="MuiTouchRipple-root-30"
  />
</button>
`;
